### PR TITLE
Adding OverFlow

### PR DESCRIPTION
**Pull request by shivammehta25**
This is the model from the paper: https://arxiv.org/abs/2211.06892
Audio samples: https://shivammehta25.github.io/OverFlow/

### CODE REVIEW

File: TTS/tts/models/overflow.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+        Returns:
+            outputs: Dictionary containing the following
+                - mel (torch.Tensor): :math:`[B, T_out, C]`
+                - hmm_outputs_len (torch.Tensor): :math:`[B]`
+                - state_travelled (List[List[int]]): List of lists containing the state travelled for each sample in the batch.
+                - input_parameters (list[torch.FloatTensor]): Input parameters to the neural HMM.
+                - output_parameters (list[torch.FloatTensor]): Output parameters to the neural HMM.
+        """
+        # TODO: inputs are
```
Comment (shivammehta25) [inline TTS/tts/models/overflow.py:200]: @erogol After the end of the epoch, while synthesising the test sentences, the inputs here are on CPU despite the model training on GPU. I am confused as to why that could be the case.

File: TTS/tts/utils/helpers.py [hunk 2]
```diff
@@ -236,3 +236,55 @@ def maximum_path_numpy(value, mask, max_neg_val=None):
     path = path * mask.astype(np.float32)
     path = torch.from_numpy(path).to(device=device, dtype=dtype)
     return path
+
+
```
Comment (erogol) [inline TTS/tts/utils/helpers.py:240]: You can place these functions into overflow.py too

**Review by erogol — CHANGES_REQUESTED** [review]

File: recipes/ljspeech/tacotron2-DDC/run-November-24-2022_01+19PM-c50d89fc/train_tacotron_ddc.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
+import os
```
[the remark this answers is not in the document]
Comment (shivammehta25) [reply recipes/ljspeech/tacotron2-DDC/run-November-24-2022_01+19PM-c50d89fc/train_tacotron_ddc.py:1]: oh yeah, my bad! I will fix this.

File: TTS/tts/models/overflow.py (hunk 1 above)
Comment (erogol) [reply TTS/tts/models/overflow.py:203]: It uses ```BaseTTS.test_run()``` if you want to debug what is wrong. It should be passing tensors to GPU normally.

Comment (shivammehta25) [reply TTS/tts/models/overflow.py:203]: I have a hunch that it could be because of `nn.Parameters(required_grad=False)` and somehow the Trainer is not putting this parameter on the GPU correctly. I am trying to replace `nn.Parameters` with `register_buffer` and see if the problem disappears.

Comment (shivammehta25) [reply TTS/tts/models/overflow.py:203]: yeap! That was it!

**Comment by king-dahmanus** [comment]
Good idea, from the samples I've hird the thing is quite good. What about adding neural hmm, or is it the same thing but just upgraded?

**Comment by king-dahmanus** [comment]
That's nice. How fast is it on the cpu? I previously suggested improving its speed for screen readers but I didn't realize how foolish that was untill recently. So how fast, or faster/slower is it compared to tacotron2 with hifigan or vits?

**Comment by shivammehta25** [comment]
These both are autoregressive models so they would be similar to Tacoron 2 which is slower than parallel models but yields better overall prosody, also OverFlow is probabilistic thus it synthesises different realisations of the same text as it maximises the true data likelihood unlike Tacotron 2 that minimises a Mean square error and generates average prosody (which is not bad either).

**Comment by king-dahmanus** [comment]
does it varry the inflection each time the same text is requested like tacotron?

File: recipes/ljspeech/tacotron2-DDC/train_tacotron_ddc.py [hunk 4]
```diff
@@ -16,7 +16,7 @@
```
[the remark this answers is not in the document]
Comment (shivammehta25) [reply recipes/ljspeech/tacotron2-DDC/train_tacotron_ddc.py]: reverted!

**Comment by erogol** [comment]
Cool the PR is ready. I'll first try the LJSpeech recipe and let you know how it goes.

File: TTS/tts/layers/overflow/neural_hmm.py [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+    Paper abstract::
+        Neural sequence-to-sequence TTS has achieved significantly better output quality than statistical speech synthesis using
+        HMMs. However, neural TTS is generally not probabilistic and uses non-monotonic attention. Attention failures increase
+        training time and can make synthesis babble incoherently. This paper describes how the old and new paradigms can be
+        combined to obtain the advantages of both worlds, by replacing attention in neural TTS with an autoregressive left-right
+        no-skip hidden Markov model defined by a neural network. Based on this proposal, we modify Tacotron 2 to obtain an
+        HMM-based neural TTS model with monotonic alignment, trained to maximise the full sequence likelihood without
+        approximation. We also describe how to combine ideas from classical and contemporary TTS for best results. The resulting
+        example system is smaller and simpler than Tacotron 2, and learns to speak with fewer iterations and less data, whilst
+        achieving comparable naturalness prior to the post-net. Our approach also allows easy control over speaking rate.
```
Comment (erogol) [inline TTS/tts/layers/overflow/neural_hmm.py:29]: Can you explain the arguments here?

File: TTS/tts/layers/overflow/neural_hmm.py [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+        # Initialize autoregression elements
+        ar_inputs = self._add_go_token(mels)
+        h_memory, c_memory = self._init_lstm_states(batch_size, self.memory_rnn_dim, mels)
+
+        for t in range(T_max):
+
+            # Process Autoregression
+            h_memory, c_memory = self._process_ar_timestep(t, ar_inputs, h_memory, c_memory)
+            # Get mean, std and transition vector from decoder for this timestep
+            # Note: Gradient checkpointing currently doesn't works with multiple gpus inside a loop
```
Comment (erogol) [inline TTS/tts/layers/overflow/neural_hmm.py:114]: If this is a blocker to use multi-gpu we should explain this in the model docstring and docs too

Comment (erogol) [reply TTS/tts/layers/overflow/neural_hmm.py:114]: Is this a model specific issue or rooted from torch

Comment (shivammehta25) [reply TTS/tts/layers/overflow/neural_hmm.py:114]: This is a torch issue gradient checkpointing in a loop is currently not supported in DDP. It works just fine for Multi-GPU if we just turn off the flag by `use_grad_checkpointing=False` but will significantly increase memory usage while training.

File: TTS/tts/layers/overflow/neural_hmm.py (hunk 5 above)
Comment (shivammehta25) [reply TTS/tts/layers/overflow/neural_hmm.py:29]: Yes! I will also add this information to the config.

Comment (shivammehta25) [reply TTS/tts/layers/overflow/neural_hmm.py:29]: Done!

File: TTS/tts/layers/overflow/neural_hmm.py (hunk 6 above)
Comment (shivammehta25) [reply TTS/tts/layers/overflow/neural_hmm.py:130]: I have added some information please take a look and see if it needs any more explanation.

File: TTS/tts/datasets/dataset.py [hunk 7]
```diff
@@ -65,6 +65,7 @@ def __init__(
         use_noise_augment: bool = False,
         start_by_longest: bool = False,
         verbose: bool = False,
+        compute_mel_statistics: bool = False,
```
Comment (Edresson) [inline TTS/tts/datasets/dataset.py:68]: The parameter "compute_mel_statistics" is used somewhere? I didn't find the place where this parameter is used.

Comment (shivammehta25) [reply TTS/tts/datasets/dataset.py:68]: Oh yeah! I added it initially but later decided to move this inside the model as only OverFlow uses it and neural HMM uses it. I will remove it!

**Comment by shivammehta25** [comment]
Oh my bad! I clicked one too many time the request for review button. Sorry for the spam xP

**Review by Edresson — APPROVED** [review]

**Review by Edresson — APPROVED** [review]

**Comment by erogol** [comment]
@shivammehta25 how do you compute `lj_parameters.pt`?

**Review by erogol — APPROVED** [review]

**Comment by shivammehta25** [comment]
Inside `TTS/tts/layers/overflow/comon_layers.py`. There exists `OverflowUtils.get_data_parameters_for_flat_start` which is called in `on_init_start` of the model. It loads the training data and computes the `means and std` (a single scalar) and for the whole training set to standardise the data during training (the model has mean and std as `registered_buffers`). This is used during the initialization in `OutputNet`, where the weights of the last layer are set to zeros and bias is set to 0 and 1 (a cool hack from fixup initialization). It also computes an average transition probability to start with an "ideal diagonal alignment". I sort of "cache" these lj_parameters.pt them near the cached phonemes.

**Comment by erogol** [comment]
Ok thanks. I missed it for some reason. I also trained an LJSpeech model. It works great. 

But I think we also need to train a vocoder that we can do separately. 

Mergin it now 👍

**Comment by shivammehta25** [comment]
I tried synthesising waveforms with the universal hifigan vocoder it works pretty well!
